pkg/provider/pulumi/gcp/policy.go

```diff
@@ -0,0 +1,281 @@
+		"secretmanager.secrets.getIamPolicy",
+		"secretmanager.version.get",
+		"secretmanager.secrets.list",
+		"secretmanager.versions.list",
+	},
+	v1.Action_SecretPut: {
+		"secretmanager.versions.add",
+	},
+}
+
+var collectionActions []string = nil
+
+func getCollectionActions() []string {
+	if collectionActions == nil {
+		collectionActions = make([]string, 0)
+		collectionActions = append(collectionActions, gcpActionsMap[v1.Action_CollectionDocumentRead]...)
+		collectionActions = append(collectionActions, gcpActionsMap[v1.Action_CollectionDocumentWrite]...)
+		collectionActions = append(collectionActions, gcpActionsMap[v1.Action_CollectionDocumentDelete]...)
+	}
+
+	return collectionActions
+}
+
+func filterCollectionActions(actions pulumi.StringArray) pulumi.StringArrayOutput {
+	arr, _ := actions.ToStringArrayOutput().ApplyT(func(actions []string) []string {
+		filteredActions := []string{}
+
+		for _, a := range actions {
+			for _, ca := range getCollectionActions() {
+				if a == ca {
+					filteredActions = append(filteredActions, a)
+					break
+				}
+			}
+		}
+
+		return filteredActions
+	}).(pulumi.StringArrayOutput)
+
+	return arr
+}
+
+func actionsToGcpActions(actions []v1.Action) pulumi.StringArray {
+	gcpActions := make(pulumi.StringArray, 0)
+
+	for _, a := range actions {
+		for _, ga := range gcpActionsMap[a] {
+			gcpActions = append(gcpActions, pulumi.String(ga))
+		}
+	}
+
+	return gcpActions
+}
+
+// Custom roles soft delete, so must have more randomization for each name, or there will be conflict on each deploy
+func newCustomRoleName(princName string) string {
+	id := uuid.New()
+	return fmt.Sprintf("%s-%s", princName, id.String())
+}
+
+func newPolicy(ctx *pulumi.Context, name string, args *PolicyArgs, opts ...pulumi.ResourceOption) (*Policy, error) {
+	res := &Policy{Name: name, RolePolicies: make([]*projects.IAMMember, 0)}
+	err := ctx.RegisterComponentResource("nitric:func:GCPPolicy", name, res, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	actions := actionsToGcpActions(args.Policy.Actions)
+
+	rolePolicy, err := projects.NewIAMCustomRole(ctx, name, &projects.IAMCustomRoleArgs{
+		Title:       pulumi.String(name),
+		Permissions: actions,
+		RoleId:      pulumi.String(strcase.ToCamel(name)),
+	}, pulumi.Parent(res))
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, principal := range args.Policy.Principals {
+		sa := args.Principals[v1.ResourceType_Function][principal.Name]
+		name := newCustomRoleName(principal.Name)
+
+		for _, resource := range args.Policy.Resources {
+			memberName := fmt.Sprintf("%s-%s", principal.Name, resource.Name)
+			memberId := pulumi.Sprintf("serviceAccount:%s", sa.Email)
+
+			switch resource.Type {
+			case v1.ResourceType_Bucket:
+				b := args.Resources.Buckets[resource.Name]
+
+				storage.NewBucketIAMMember(ctx, memberName, &storage.BucketIAMMemberArgs{
+					Bucket: b.Name,
+					Member: memberId,
+					Role:   rolePolicy.Name,
+				}, pulumi.Parent(res))
```

Comment:
```suggestion
				}, append(opts, pulumi.Parent(res))...)
```
as with the others below

Reply:
I'm not sure we want to propogate resource options to all child resources, but apply them to the parent resource instead.

Reply:
ok, well some of the providers do this. so we should at least be consistent one way or the other.